lib/chef/knife/bootstrap.rb

```diff
@@ -638,8 +638,10 @@ def connect!
           EOM
           # FIXME: this should save the key to known_hosts but doesn't appear to be
           config[:ssh_verify_host_key] = :accept_new
-          do_connect(connection_opts(reset: true))
-        elsif ssh? && e.cause && e.cause.class == Net::SSH::AuthenticationFailed
+          conn_opts = connection_opts(reset: true)
+          opts.merge! conn_opts
```

Comment:
Why the change to merge into old opts here?

Reply:
Because if --connection-password is not passed and `Net::SSH::AuthenticationFailed` or `Train::ClientError` error is thrown then old opts will have the password.